### ISSUE TITLE
StopSpam/StopSpamMod: Jabber message query overflow

### ISSUE THREAD

**Issue by dartraiden** [miranda-ng/miranda-ng#3679]
When a message is received, StopSpam suppresses the `auth req` creation, sends a security question, and waits for the correct answer. The original message is placed in the offline message queue on the server since there was no response to the  `auth req`. Thus, each message from the spammer is added to the queue. Sooner or later, the queue becomes full and the user cannot receive offline messages.

**Comment by dartraiden** [comment]
Кажется, я придумал решение проблемы. Опишу её подробно:

# Проблема

jabber.ru (сервер) настроен так, что не позволяет писать тем, кто тебя не авторизовал. Таким образом, спамеры, которых за день у меня бывает по 10 и более, сначала запрашивают авторизацию. Если ты дал авторизацию, они тебе следом засылают уже свою спам-мессагу., Естественно, я не хочу им авторизацию давать. Но постоянно видеть запросы авторизации и руками жать Deny это гемор

# Имеющиеся способы борьбы
**Все существующие в Миранде способы борьбы с этим не эффективны**. Рассмотрим их и их недостатки.

## Способ 1: Средствами самой Миранды (Контакты->Игнорировать) скрыть все запросы авторизации от "Неизвестных контактов", а также в корневой группе скрыть тех, у кого нет группы.

Спамер засылает запрос авторизации. Миранда его не показывает. Спамер влетает в базу в корневую групс `NotOnList`, в клисте он не показывается (т.к. группы у него нет), а при выключении Миранды плагин клиста мочит всех, кто не в ростере.

Недостатки:
- если ты не ответил на запрос авторизации (положительно или отрицательно), то этот запрос остаётся висеть на сервере. При следующем выходе в онлайн все эти запросы сервер пришлёт. Их обработка (их надо принять, скрыть, завести в базе контакт) занимает время. С каждым днём эта очередь неотвеченных авторизаций всё длиннее. В итоге, когда их число достигает порядка сотни, Миранда после старта где-то минуту не может делать ничего: ни принимать, ни отправлять сообщения, потому что поток занят обработкой этого мусора.
- неотвеченные запросы авторизации забивают серверную очередь сообщений. На jabber.ru это в районе сотни. Как только у тебя набралась сотня таких проигнорированных спамеров, тебе добросовестные контакты из ростера не могут написать в оффлайн. Потому что у тебя на сервере кончилось место под оффлайновые сообщения.

## Способ 2: Отбивать запросы авторизации антиспамом и им же скрывать неответивших

По сревнение со способом 1 этот способ имеет достоинства:
- не-спамеры имеют возможность пройти вопрос антиспама и заслать тебе авторизацию
- не нужно включить скрытие контактов в корневой группе, им антиспам вешает `Hidden`

Но недостатки у него те же:
- он не отказывает в авториации спамерам, а просто скрывает их запросы
- очередь сообщений на сервере переполняется

**Comment by dartraiden** [comment]
# Предлагаемое решение

Если спамер не прошёл вопрос антиспама, то запрос авторизации **не скрывать, а отказывать в ней**.

**Comment by dartraiden** [comment]
1) До фикса если включена опция "блокировать авторизацию", то в ответ на запрос авторизации отправлялся вопрос антиспама, чтобы человек мог его пройти. Это было правильно.

 После фикса вопрос не отправляется, через 5 минут контакт получает отказ в авторизации и стирается из базы. Таким образом, теперь эта опция вообще лишает нормального человека возможности пройти антиспам. Это неправильно.

2) Перестала работать опция "Записывать в системную историю".

**Comment by georgehazan** [comment]
2. раз в минуту таймер проверяет протухшие контакты, для которых запрошено стирание.

**Comment by dartraiden** [comment]
Да, теперь всё ок. Я неправильно понимал работу опции "Блокировать запросы авторизации". Я думал, что эта опция распространяет работу антиспама на авторизацию. В реальности эта опция тупо скрывала запросы авторизации.
